conu/backend/docker/container.py

```diff
@@ -8,49 +8,112 @@
+        Build `docker run` command
+
+        :param command: list of str, command to run in the container, examples:
+            - ["ls", "/"]
+            - ["bash", "-c", "ls / | grep bin"]
+        :param additional_opts: list of str, additional options for `docker run`
+        """
+        self.binary = ["docker"]
+        self.global_options = []
+        self.command = ["container", "run"]
+        self.options = [] + (additional_opts or [])
+        self.image_name = None
+        self.arguments = command or []
+
+    def __str__(self):
+        return str(self.build())
+
+    def build(self):
+        return self.binary + self.global_options + self.command + self.options + \
+            [self.image_name] + self.arguments
+
+
 class DockerContainer(Container):
-    def __init__(self, image, tag=None):
-        super(DockerContainer, self).__init__(image, None)
-        self.tag = tag or random_str()
+    def __init__(self, image, container_id, name=None, popen_instance=None):
+        """
+        :param image: DockerImage instance
+        :param container_id: str, unique identifier of this container
+        :param name: str, pretty container name
+        :param popen_instance: instance of Popen (if container was created using method
+            `via_binary`, this is the docker client process)
+        """
+        super(DockerContainer, self).__init__(image, container_id, name)
+        self.popen_instance = popen_instance
         if not isinstance(image, DockerImage):
-            raise BaseException("Image is not instance of Image class")
-        self.json = None
-        self.image = image
-        self.docker_id = None
-        self.__occupied = False
-
-    def get_tag_name(self):
-        return self.tag
-
-    def inspect(self, force=False):
-        if force or not self.json:
-            output = json.loads(run_cmd("docker container inspect %s" % self.tag))[0]
-            self.json = output
-            return output
-        else:
-            return self.json
-
-    def check_running(self):
-        i = self.inspect(force=True)
+            raise RuntimeError("image argument is not an instance of the DockerImage class")
+
+    def __repr__(self):
+        return "DockerContainer(image=%s, id=%s)" % (self.image, self.get_id())
 
+    def __str__(self):
+        return self.get_id()
+
+    def get_id(self):
+        """
+        get unique identifier of this container
+
+        :return: str
+        """
+        if self._id is None:
+            # FIXME: provide a better error message when key is not defined
+            self._id = self.get_metadata(refresh=False)["Id"]
+        return self._id
+
+    def inspect(self, refresh=False):
+        """
+        return cached metadata by default (a convenience method)
+
+        :param refresh: bool, returns up to date metadata if set to True
+        :return: dict
+        """
+        return self.get_metadata(refresh=refresh)
+
+    def get_metadata(self, refresh=False):
```

Comment:
Would it make more sense to have `refresh=True` by default? IMHO most of the time you want to know the up-to-date information at the time of asking, but that may just be my feeling.

Reply:
I set it to `False` because that's what I have in sen, the issue over there is that it takes quite some time to do `inspect` and all containers and images -- I think we won't run into that here, will change!